### DISCUSSION TITLE
Full-stack application: task `client:dev` runs but `:dev` fails

### DISCUSSION THREAD

**Discussion by diraneyya** [moonrepo/moon#830] [Q&A] [answered]
This is the issue created by one of my students: https://github.com/WildCodeSchool/2023-01-EN-Berlin-Remote2-Project3/issues/2

And this is our mono-repo for the final project: https://github.com/WildCodeSchool/2023-01-EN-Berlin-Remote2-Project3

**Steps to reproduce:**
- clone the repo
- run the command `moon run client:dev` (the Vite development server runs)
- run the command `moon run :dev` (the backend runs but the Vite development server fails)

What am I missing?

**Comment by milesj** [comment]
@diraneyya What version of moon is being used? It's not pinned in the repo. Does this happen on the latest?

**Reply by diraneyya** [reply]
Yes I am using the latest version (1.4.0).

**Comment by milesj** [comment]
Ok, after looking at the repo, unfortunately this is hitting an edge case in the dep graph where the `client:dev` and `server:dev` tasks end up in different "groups". Since the client group comes after the server group, it never runs because server never "completes".

If we look at the dep graph, you can see server is far more complex:
<img width="790" alt="Screenshot 2023-05-05 at 12 50 48 PM" src="https://user-images.githubusercontent.com/143744/236556186-4f334c8f-c570-4a7a-a9af-6f2429420e9a.png">

This results in the following groups:
```
[
    [
        NodeIndex(5), // RunTarget(server:dev)
        NodeIndex(3), // RunTarget(client:dev)
    ],
    [
        NodeIndex(8), // RunTarget(server:dev-start)
        NodeIndex(6), // RunTarget(server:dev-build)
        NodeIndex(2), // SyncNodeProject(client)
    ],
    [
        NodeIndex(7), // RunTarget(server:generate-prisma-client)
    ],
    [
        NodeIndex(4), // SyncNodeProject(server)
        NodeIndex(1), // InstallNodeDeps(18.15.0)
    ],
    [
        NodeIndex(0), // SetupNodeTool(18.15.0)
    ],
]
```

Unfortunately there's no way to "fix" the graph at this time, besides running in separate terminals. You can try creating a master `dev-all` task, that has `client:dev` and `server:dev` as deps, but I'm not 100% sure it will work.

This is something we're hoping to fix in dep graph v2. I'll take a look at it again for an upcoming release.

**Comment by milesj** [comment]
This should be fixed in v1.6: https://github.com/moonrepo/moon/pull/848

**Reply by diraneyya** [reply, marked as answer]
Thank you!

In the meantime I have [fixed](https://github.com/WildCodeSchool/2023-01-EN-Berlin-Remote2-Project3/commit/37279b4ccf818df65609ebdc5161e7afcbf38220) this issue by changing the task from a `noop` task with dependencies to a task that runs the command: `moon run server:dev-build server:dev-start`. It is less elegant and it contains a redundant reference of the project `self` but it works.

I suppose that this is equivalent or _along the lines_ of your suggestion of:
> creating a master dev-all task, that has client:dev and server:dev as deps

https://github.com/WildCodeSchool/2023-01-EN-Berlin-Remote2-Project3/pull/5

**Comment by diraneyya** [comment]
We upgraded to moon 1.6 and this now works!

Link to the commit upgrading: https://github.com/WildCodeSchool/2023-01-EN-Berlin-Remote2-Project3/commit/9cae303a7a89aa3873e7b18d74b9266453e14250

As a side note that my previous suggested fix (marked as solution) no longer works in moon 1.6+. Namely, trying to use the `moon run ...` command in the command field of a task as a workaround to running multiple tasks without marking them as deps would lead to the following cryptic error message:
```
moon run :dev
▪▪▪▪ server:dev
server:dev | error: invalid value 'moon=info,proto=info,starbase=info' for '--log <LOG>'
server:dev |   [possible values: off, error, warn, info, debug, trace]
server:dev | 
server:dev | For more information, try '--help'.
▪▪▪▪ server:dev (4ms, adc1b3a7)
[ERROR 11:38:14] log  Encountered a critical error, aborting the action pipeline  log.target="moon:action-pipeline:batch:4" log.module_path="moon_action_pipeline::pipeline" log.file="crates/core/action-pipeline/src/pipeline.rs" log.line=195

 ERROR  Process moon failed with a 2 exit code.
```

**Reply by milesj** [reply]
That's an interesting one 🤔
